### ISSUE TITLE
FR: Sliders for better highlighting and selecting of words

### ISSUE THREAD

**Issue by TheWranger** [koreader/koreader#9653]
**Does your feature request involve difficulty completing a task? Please describe.**
When trying to highlight and select a word, it is sometimes hard to see if it's been selected as your finger is covering the word and you cannot see the highlight. It is also unintuitive when trying to expand a highlight due to it being hard to judge whether the existing selection has been "picked up" by your touch and you can drag to expand it. Sometimes I hold down on a selection then drag to expand but nothing happens as I haven't held down for long enough or e-ink hasn't detected it. The alternative to just tap on an existing selection and press the buttons to manually increase the selection word-by-word is also quite slow.

**Describe the solution you'd like**
Steal design idea of using two sliders beside selected words. When you tap and hold on a word in an epub, alongside it being highlighted two sliders will show up around it. This fixes the problem of not being able to see if a word's been highlighted or selected due to it being under your finger.  

You can press and hold on the sliders, and in doing so they will turn to black and then you can drag them to expand your highlight. With them turning black, this fixes the problem of not being able to initially tell if you've "grabbed" the highlight or not - there is now a visual response.

**In summary**, I think a non-intrusive solution would be:
- Add an option to settings, or perhaps make this its own separate toggleable plugin, to alter the behaviour of "selecting" text (when you tap on "select" in the long-press popup dialog) so that handles or sliders will appear beside it which can be long-pressed on and dragged to expand the selection of text. This would hijack/replace the current behaviour of asking to "select ending fragment".
- The sliders should also appear when pressing on an existing highlight (when the popup to "delete", "style", "add note" appears), so that they can be long-pressed and dragged to easily expand the highlight. The existing popup should still continue to appear alongside the sliders, but when the sliders are being pressed and dragged, the popup should disappear to allow better viewing of the document. (This also solves a current problem with not being able to see the text underneath the popup when expanding the highlight).
- Some part of the sliders should turn black, or some shade, when they are being long-pressed and dragged (like how Kobo handles it) - as this gives visual feedback to the reader

**Describe alternatives you've considered**
I haven't considered any alternatives.

**Additional context**

Here are some visual examples of sliders in use:
Android, Kobo Books:
![image](https://user-images.githubusercontent.com/24642838/196951550-8d5fca29-0705-4ddf-8637-4a7fa8fa6465.png)

iOS, Kobo Books:
![image](https://user-images.githubusercontent.com/24642838/196951706-8abf0133-646d-463e-9c8a-bf46c4cedc15.png)

Kindle:
![image](https://user-images.githubusercontent.com/24642838/196952127-1acf92d7-dad6-427c-acee-3e8fa65f695a.png)

Kobo:
![image](https://user-images.githubusercontent.com/24642838/197114049-a350be24-0bdf-4ca3-90b7-deb2d2f7e6ce.png)

**Comment by hius07** [comment]
One more alternative way currently available
- in the "Long-press on text" disable "Dictionary on sigle word selection" and select "Ask with popup dialog"
- long-press on the first word of the fragment to be selected
- in the popup dialog press Select button
- long-press on the last word of the fragment

**Comment by NiLuJe** [comment]
We had a discussion about this a while back, IIRC (possibly a *long* way back, though :D).

(Possible in the issue where we implemented the selection update popup?).

(To recap: I'm... not a fan of handles, doing it right is harder than it looks, and most the time they just clutter the screen without really helping. I quite like what Plato does here, even if I don't even remember what that is right now :D).

**Comment by TheWranger** [comment]
> One more alternative way currently available
> 
> * in the "Long-press on text" disable "Dictionary on sigle word selection" and select "Ask with popup dialog"
> * long-press on the first word of the fragment to be selected
> * in the popup dialog press Select button
> * long-press on the last word of the fragment

Yeah, I overlooked this. This way does avoid some of the problem of long-pressing and e-ink missing your selection as you drag your finger. I've noticed that it's still very easy, when doing this, to miss something like a period at the end of the sentence, because your finger is blocking your sight on what your pressing down on. Then when using the highlight expander menu (this thing in the pic below) to try and expand the highlight slightly, it'll pick up the ending period, but also the word in the sentence afterwards so you aren't able to *only* get the period.
![image](https://user-images.githubusercontent.com/24642838/197110131-f8ecb9c2-1d4c-4178-bec4-c1f368986477.png)


> We had a discussion about this a while back, IIRC (possibly a _long_ way back, though :D).
> 
> (Possibly in the issue where we implemented the selection update popup?).
> 
> (To recap: I'm... not a fan of handles, doing it right is harder than it looks, and most the time they just clutter the screen without really helping (keep in mind I have tiny fingers, which quite possibly affects my view on the matter). I quite like what Plato does for selections though, even if I don't even remember what that is right now :D).

I'll have to try out Plato and see how they do it

**Comment by TheWranger** [comment]
I cleaned up the OP and added a summary of the solution I think fits best into KOReader. I also wanted to put a $10 bounty on this, but it seems like BountySource is quite broken at the moment - can't sign up and API seems to be down for over a month. :(

**Comment by mergen3107** [comment]
If you tap and hold these arrows you can highlight character by character, including periods.

**Comment by poire-z** [comment]
> when using the highlight expander menu (this thing in the pic below) to try and expand the highlight slightly, it'll pick up the ending period, but also the word in the sentence afterwards so you aren't able to only get the period.

Tap on the `###>` buttons will indeed grab by word. You can long-press on them to grab by character one by one (so, picking only your period).

> I find it more comfortable to use my thumb to highlight :(

Indeed, use the right tool for the job (the pinky finger if you should :), highlighting with a hammer ought to be hard :)

Also not a fan on handles, and probably hard to implement the way we do highlights (ie; with crengine/EPUBs, it's the engine that does the highlighting, we would have to have frontend draw stuff over...)

**Comment by uncharted-user** [comment]
Is is correct that Bookmark with Highlight sync is not available for multiple devices in KOReader similar to Kindle app? I am aware of KOHighlight, Progress sync, Export highlights.

**Comment by offset-torque** [comment]
> **TheWranger:** Yeah, I overlooked this. This way does avoid some of the problem of long-pressing and e-ink missing your selection as you drag your finger. I've noticed that it's still very easy, when doing this, **to miss something like a period at the end of the sentence, because your finger is blocking your sight on what your pressing down on.**

I think **poire-z** fixed this problem (thanks!). Currently I am using this highlighting style and it automatically adds punctuation like commas and periods to the highlight coming after the last word. Even quotation marks are added correctly to the both ends.

Is this not working for you ?

**Comment by TheWranger** [comment]
> > **TheWranger:** Yeah, I overlooked this. This way does avoid some of the problem of long-pressing and e-ink missing your selection as you drag your finger. I've noticed that it's still very easy, when doing this, **to miss something like a period at the end of the sentence, because your finger is blocking your sight on what your pressing down on.**
> 
> I think **poire-z** fixed this problem (thanks!). Currently I am using this highlighting style and it automatically adds punctuation like commas and periods to the highlight coming after the last word. Even quotation marks are added correctly to the both ends.
> 
> Is this not working for you ?

It's inconsistent. It seems like, for larger highlights (containing several paragraphs), the ending punctuation will not be automatically picked up. Bearing in mind that I'm using dev version v2022.09-97-g60a16c73_2022-10-21 though, so maybe it works fine on the previous stable. I will test it some more and try to cite a consistent example from a book off gutenberg. I'm on running the dev version as I'm using a Clara 2E at the moment

**Comment by TheWranger** [comment]
> > **TheWranger:** Yeah, I overlooked this. This way does avoid some of the problem of long-pressing and e-ink missing your selection as you drag your finger. I've noticed that it's still very easy, when doing this, **to miss something like a period at the end of the sentence, because your finger is blocking your sight on what your pressing down on.**
> 
> I think **poire-z** fixed this problem (thanks!). Currently I am using this highlighting style and it automatically adds punctuation like commas and periods to the highlight coming after the last word. Even quotation marks are added correctly to the both ends.
> 
> Is this not working for you ?

Can be reproduced on Stable 2022.10. Here you can see it in Bartleby the Scrivener
![image](https://user-images.githubusercontent.com/24642838/197395236-b1aa8375-f02e-46f8-b60d-f0e0881196c7.png)

Can be reproduced on both the standard EPUB as well as EPUB3.
https://www.gutenberg.org/ebooks/11231

Should I make a separate issue for this, or is this good enough as it is?

**Comment by poire-z** [comment]
The behaviour (including or not the punctuation at the end of your single highlight) depends on the start of your highlight (which I guess was not a "sentence" start).
The specs :) has been discussed/described at/around https://github.com/koreader/koreader/issues/8857#issuecomment-1151198923 and implemented in #9198.

So, it might not be an issue.

**Comment by TheWranger** [comment]
I think I was wrong. When I made this request the e-reader I was using actually had a defective touch screen, and I didn't find this out until recently, so highlighting in koreader was made more unresponsive than usual and I didn't realise this wasn't the norm. The problem persisted on the stock OS (kobo's OS), but having sliders made the problem less noticeable.

Using koreader on a proper working device, and also a larger one, I think the functionality is fine as it is. 

However, I think anyone reading with a smaller font will definitely still notice that it's harder to highlight without any sliders, as it's harder to accurately "grip" words under your finger when they are so small. If it's so hard to implement then I'm not sure if it's worth worrying about over other issues, though.

> The behaviour (including or not the punctuation at the end of your single highlight) depends on the start of your highlight (which I guess was not a "sentence" start). The specs :) has been discussed/described at/around [#8857 (comment)](https://github.com/koreader/koreader/issues/8857#issuecomment-1151198923) and implemented in #9198.
> 
> So, it might not be an issue.

You're right, I was wrong.
